GVFS/GVFS/CommandLine/UpgradeVerb.cs

```diff
@@ -0,0 +1,384 @@
+using CommandLine;
+using GVFS.Common;
+using GVFS.Common.Tracing;
+using GVFS.Upgrader;
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace GVFS.CommandLine
+{
+    [Verb(UpgradeVerbName, HelpText = "Checks for new GVFS release, downloads and installs it when available.")]
+    public class UpgradeVerb : GVFSVerb.ForNoEnlistment
+    {
+        private const string UpgradeVerbName = "upgrade";
+        private ITracer tracer;
+        private ProductUpgrader upgrader;
+        private InstallerPreRunChecker prerunChecker;
+        private ProcessLauncher processLauncher;
+
+        public UpgradeVerb(
+            ProductUpgrader upgrader,
+            ITracer tracer,
+            InstallerPreRunChecker prerunChecker,
+            ProcessLauncher processWrapper,
+            TextWriter output)
+        {
+            this.upgrader = upgrader;
+            this.tracer = tracer;
+            this.prerunChecker = prerunChecker;
+            this.processLauncher = processWrapper;
+            this.Output = output;
+        }
+
+        public UpgradeVerb()
+        {
+            this.processLauncher = new ProcessLauncher();
+            this.Output = Console.Out;
+        }
+
+        [Option(
+            "confirm",
+            Default = false,
+            Required = false,
+            HelpText = "Pass in this flag to actually install the newest release")]
+        public bool Confirmed { get; set; }
+
+        protected override string VerbName
+        {
+            get { return UpgradeVerbName; }
+        }
+        
+        public override void Execute()
+        {
+            ReturnCode exitCode = ReturnCode.Success;
+            if (!this.TryInitializeUpgrader() || !this.TryRunProductUpgrade())
+            {
+                exitCode = ReturnCode.GenericError;
+                this.ReportErrorAndExit(this.tracer, exitCode, string.Empty);
+            }
+        }
+
+        private bool TryInitializeUpgrader()
+        {
+            OperatingSystem os_info = Environment.OSVersion;
+
+            if (os_info.Platform == PlatformID.Win32NT)
+            {
+                if (this.upgrader == null)
+                {
+                    JsonTracer jsonTracer = new JsonTracer(GVFSConstants.GVFSEtwProviderName, "UpgradeVerb");

```

Comment:
Here's an example of a log file from a successful upgrade run:
```
[10/4/2018 8:31:05 AM] TryCheckUpgradeInstallable (Start)
[10/4/2018 8:31:05 AM] TryRunPreUpgradeChecks (Start)
[10/4/2018 8:31:06 AM] Information {"Message":"Successfully finished pre upgrade checks. Okay to run `gvfs upgrade`."}
[10/4/2018 8:31:06 AM] TryRunPreUpgradeChecks (Stop) {"DurationMs":132}
[10/4/2018 8:31:06 AM] Information {"Message":"Upgrade is installable."}
[10/4/2018 8:31:06 AM] TryCheckUpgradeInstallable (Stop) {"DurationMs":138}
[10/4/2018 8:31:06 AM] TryCheckUpgradeAvailable (Start)
[10/4/2018 8:31:06 AM] Information {"Message":"Successfully checked server for GVFS upgrades."}
[10/4/2018 8:31:06 AM] TryCheckUpgradeAvailable (Stop) {"DurationMs":269}
[10/4/2018 8:31:06 AM] TryCopyUpgradeTool (Start)
[10/4/2018 8:31:06 AM] Information {"Message":"Successfully Copied upgrade tool to C:\\ProgramData\\GVFS\\GVFS.Upgrade\\Tools\\GVFS.Upgrader.exe"}
[10/4/2018 8:31:06 AM] TryCopyUpgradeTool (Stop) {"DurationMs":77}
[10/4/2018 8:31:06 AM] TryLaunchUpgradeTool (Start)
[10/4/2018 8:31:06 AM] Information {"Message":"Successfully launched upgrade tool."}
[10/4/2018 8:31:06 AM] TryLaunchUpgradeTool (Stop) {"DurationMs":358}
```

I can't tell from this log:
* what version we started from
* what version we went to
* if the installers had any issues or if everything installed correctly


And here's an example of a log file from a failed install:
```
[10/4/2018 8:33:21 AM] TryCheckUpgradeInstallable (Start)
[10/4/2018 8:33:21 AM] TryRunPreUpgradeChecks (Start)
[10/4/2018 8:33:23 AM] Error {"ErrorMessage":"TryRunPreUpgradeChecks: `gvfs upgrade` is not supported because you have previously installed an out of band ProjFS driver.\r\nCheck your team's documentation for how to upgrade."}
[10/4/2018 8:33:23 AM] TryRunPreUpgradeChecks (Stop) {"DurationMs":1329}
[10/4/2018 8:33:23 AM] TryCheckUpgradeInstallable (Stop) {"DurationMs":1335}
[10/4/2018 8:33:23 AM] Error {"ErrorMessage":"TryLoadUpgradeRing failed. Invalid upgrade ring `Foo` specified in gvfs config.\r\nTo set or change upgrade ring, run `gvfs config upgrade.ring [\"Fast\"|\"Slow\"|\"None\"]` from a command prompt."}
[10/4/2018 8:33:23 AM] Error {"ErrorMessage":"TryRunProductUpgrade: Could not load upgrade ring. Invalid upgrade ring `Foo` specified in gvfs config.\r\nTo set or change upgrade ring, run `gvfs config upgrade.ring [\"Fast\"|\"Slow\"|\"None\"]` from a command prompt."}
```

The problem with this is that this is the exact same info that was shown to the user. So the log files aren't actually giving us any extra diagnostic info right now.

Reply:
I can see that the other log file from the upgrade tool has much more useful info in it:
```
[10/4/2018 8:31:07 AM] TryCheckIfUpgradeAvailable (Start)
[10/4/2018 8:31:07 AM] Information {"Message":"Successfully checked for new release. 1.0.18234.4"}
[10/4/2018 8:31:07 AM] TryCheckIfUpgradeAvailable (Stop) {"DurationMs":389}
[10/4/2018 8:31:07 AM] TryGetNewGitVersion (Start)
[10/4/2018 8:31:07 AM] Information {"Message":"Successfully read Git version 2.18.0.gvfs.1.26"}
[10/4/2018 8:31:07 AM] TryGetNewGitVersion (Stop) {"DurationMs":2}
[10/4/2018 8:31:07 AM] Installed Version {"installedGVFSVersion":"0.2.173.2","installedGitVersion":"2.18.0.gvfs.1.26"}
[10/4/2018 8:31:07 AM] Available Version {"gvfsVersion":"1.0.18234.4","gitVersion":"2.18.0.gvfs.1.26"}
[10/4/2018 8:31:07 AM] TryRunPreUpgradeChecks (Start)
[10/4/2018 8:31:07 AM] Information {"Message":"Successfully finished pre upgrade checks. Okay to run `gvfs upgrade`."}
[10/4/2018 8:31:07 AM] TryRunPreUpgradeChecks (Stop) {"DurationMs":5}
[10/4/2018 8:31:07 AM] TryDownloadUpgrade(1.0.18234.4) (Start)
[10/4/2018 8:31:13 AM] Information {"Message":"Successfully downloaded version: 1.0.18234.4"}
[10/4/2018 8:31:13 AM] TryDownloadUpgrade(1.0.18234.4) (Stop) {"DurationMs":5945}
[10/4/2018 8:31:13 AM] Information {"Message":"Unmounting any mounted GVFS repositories."}
[10/4/2018 8:31:13 AM] TryUnmountAllGVFSRepos (Start)
[10/4/2018 8:31:14 AM] Information {"Message":"Checking if GVFS or dependent processes are running."}
[10/4/2018 8:31:15 AM] Information {"Message":"Successfully unmounted repositories."}
[10/4/2018 8:31:15 AM] TryUnmountAllGVFSRepos (Stop) {"DurationMs":1344}
[10/4/2018 8:31:15 AM] TryInstallGitUpgrade(2.18.0.gvfs.1.26) (Start)
[10/4/2018 8:31:47 AM] Information {"Message":"Successfully installed Git version: 2.18.0.gvfs.1.26"}
[10/4/2018 8:31:47 AM] TryInstallGitUpgrade(2.18.0.gvfs.1.26) (Stop) {"DurationMs":32863}
[10/4/2018 8:31:47 AM] TryInstallGVFSUpgrade(1.0.18234.4) (Start)
[10/4/2018 8:32:10 AM] Information {"Message":"Successfully installed GVFS version: 1.0.18234.4"}
[10/4/2018 8:32:10 AM] TryInstallGVFSUpgrade(1.0.18234.4) (Stop) {"DurationMs":22052}
[10/4/2018 8:32:10 AM] Newly Installed Version {"gvfsVersion":"1.0.18234.4","gitVersion":"2.18.0.gvfs.1.26"}
```

So unless we think we can add some useful diagnostic info to the first log file, I would just drop it.

(And the upgrade tool's log file still needs to add what version we're starting from)

Reply:
In the same vein, gvfs diagnose does not capture the upgrade logs. Should it @sanoursa ?

Reply:
Yes it captured the logs for me. It's in the `GVFS.Upgrade` folder in the top level of the zip file.

Reply:
@sanoursa pointed out i upgraded to a public version that doesn't support this -- please disregard.

Reply:
These are the errors that can get logged into this file
1. FileBasedCollectionException info when trying to load ring config.
2. Network exception while trying to fetch Github/release JSON.
3. Parse errors while parsing Release JSON.
4. File system exceptions while trying to copy upgrader tool to temp directory in ProgramData.
5. Process exceptions while trying to launch Upgrader tool.

These errors don’t happen normally, thats the reason the log file looks un-eventful and more like a debug trace log. Its still good to keep this log file, b’coz if for some reason Upgrader tool didn’t get launched, or upgrade check failed, this is the file where we would look, for the detailed exception info.

Example: I simulated a file copy error and this is what the error log looked like
```
[10/4/2018 4:32:33 PM] TryCheckUpgradeInstallable (Start)
[10/4/2018 4:32:33 PM] TryRunPreUpgradeChecks (Start)
[10/4/2018 4:32:35 PM] Information {"Message":"Successfully finished pre upgrade checks. Okay to run `gvfs upgrade`."}
[10/4/2018 4:32:35 PM] TryRunPreUpgradeChecks (Stop) {"DurationMs":1820}
[10/4/2018 4:32:35 PM] Information {"Message":"Upgrade is installable."}
[10/4/2018 4:32:35 PM] TryCheckUpgradeInstallable (Stop) {"DurationMs":1825}
[10/4/2018 4:32:35 PM] TryCheckUpgradeAvailable (Start)
[10/4/2018 4:32:35 PM] Information {"Message":"Successfully checked server for GVFS upgrades."}
[10/4/2018 4:32:35 PM] TryCheckUpgradeAvailable (Stop) {"DurationMs":450}
[10/4/2018 4:32:35 PM] TryCopyUpgradeTool (Start)
[10/4/2018 4:32:35 PM] Error {"Method":"TrySetupToolsDirectory","Exception":"System.IO.FileNotFoundException: Could not find file 'M:\\Work\\GVFS\\BuildOutput\\GVFS.Windows\\bin\\x64\\Debug\\GVFS.Upgrader.exe'.\r\nFile name: 'M:\\Work\\GVFS\\BuildOutput\\GVFS.Windows\\bin\\x64\\Debug\\GVFS.Upgrader.exe'\r\n   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)\r\n   at System.IO.File.InternalCopy(String sourceFileName, String destFileName, Boolean overwrite, Boolean checkHost)\r\n   at GVFS.Common.ProductUpgrader.TrySetupToolsDirectory(String& upgraderToolPath, String& error)","ErrorMessage":"Error copying M:\\Work\\GVFS\\BuildOutput\\GVFS.Windows\\bin\\x64\\Debug\\GVFS.Upgrader.exe to C:\\ProgramData\\GVFS\\GVFS.Upgrade\\Tools\\GVFS.Upgrader.exe."}
[10/4/2018 4:32:35 PM] TryCopyUpgradeTool (Stop) {"DurationMs":5}
[10/4/2018 4:32:35 PM] Error {"ErrorMessage":"TryRunProductUpgrade: Could not launch upgrade tool. File copy error - Could not find file 'M:\\Work\\GVFS\\BuildOutput\\GVFS.Windows\\bin\\x64\\Debug\\GVFS.Upgrader.exe'."}
```